### PR TITLE
MRG: BEM model and solution in Python

### LINKED ISSUE #2190: implement bem-sol.fif code in python
for @Eric89GXL ...

### PR DESCRIPTION
**Pull request by Eric89GXL**
Closes #2190.

~~I translated all the code but haven't tested it at all. In fact I know it will not work. But the current plan for usage is something like this:~~
```
from mne import make_bem_model, make_bem_solution
subject = 'sample'
model = make_bem_model(subject)  # puts surfaces and conductances together
solution = make_bem_solution(model)  # does the potential computations
```

### CODE REVIEW

File: mne/bem.py [hunk 1]
```diff
@@ -6,14 +6,457 @@
...
+    y1 = to_rr[to_tri[0]] - fro
+    y2 = to_rr[to_tri[1]] - fro
+    y3 = to_rr[to_tri[2]] - fro
+    # This circularity makes things easy for us...
+    yy = [y1, y2, y3, y1, y3]
+    cross = np.cross(y1, y2)
+    triple = np.dot(cross, y3)
+
+    # XXX refactor with _get_solids?
+    l1 = np.sqrt((y1 * y1).sum())
```
Comment (agramfort) [inline mne/bem.py]: (y1 * y1).sum() -> np.dot(y1, y1)

fix below too.

Comment (agramfort) [inline mne/bem.py]: Also you use use functions from math module for floats: sqrt, abs ...

Comment (Eric89GXL) [inline mne/bem.py]: yeah, the reason I have it in this form is that I am almost certainly going to vectorize the operation, in which case having it in the `sum` form will be more convenient to start translating to higher dimensions

**Comment by Eric89GXL** [comment]
It seems to be working for me now, but it is horribly slow. Hopefully I can speed it up tomorrow.

**Comment by dgwakeman** [comment]
You are awesome! I am so excited for this!

**Comment by Eric89GXL** [comment]
FYI I added a couple more files to the testing dataset. BEMs have now been created for ico-2 surfaces (320 instead of 5120 or 1280) which is much faster.

The code at this point seems to work for homog 1-layer, I just need to track down a bug in the 3-layer case. Speed isn't too much worse than C for small shells (<= 1280) and actually faster for standard size (5120) because, while calculating the matrix coefficients is slower in Python, the inversion is much faster:
```
ico       type    layers  time
2 (320)   C       1       0.0300931930542
2 (320)   Python  1       0.250679969788
2 (320)   C       3       0.347531795502
2 (320)   Python  3       2.41311502457
3 (1280)  C       1       0.621548891068
3 (1280)  Python  1       2.16235113144
3 (1280)  C       3       12.2742669582
3 (1280)  Python  3       21.4681899548
4 (5120)  C       1       24.4842910767
4 (5120)  Python  1       27.2853899002
4 (5120)  C       3       643.578150034
4 (5120)  Python  3       282.933717012
```
Just need to track down where that bug is, which I probably won't get to until next week.

**Comment by Eric89GXL** [comment]
Okay, 1- and 3-layer model and solution computation works now. Ready for review/merge from my end.

**Comment by Eric89GXL** [comment]
Ping @agramfort @dengemann this is ready for review

**Comment by dengemann** [comment]
@Eric89GXL amazing! How fast is it compared to the original code?

**Comment by Eric89GXL** [comment]
See comment above, faster for normal use case (3-layer 5120)

**Comment by Eric89GXL** [comment]
There might be additional ways of optimizing it, too, if there is interest

**Comment by dengemann** [comment]
> here might be additional ways of optimizing it, too, if there is interest

I'm already very impressed. Eric what do we need here, any testing desired?

**Comment by Eric89GXL** [comment]
Hopefully the unit tests are sufficient to make us confident in using it. But sure, if you want to test it out and make comments, it's welcome.

**Comment by dengemann** [comment]
> But sure, if you want to test it out and make comments, it's welcome.

I will trust you if you're confident ;)

File: mne/bem.py [hunk 2]
```diff
@@ -12,10 +12,512 @@
...
+
+# The following approach is based on:
+#
+# de Munck JC: "A linear discretization of the volume conductor boundary
+# integral equation using analytically integrated elements",
+# IEEE Trans Biomed Eng. 1992 39(9) : 986 - 990
+#
+
+
+def _calc_beta(rk, rk1):
```
Comment (dengemann) [inline mne/bem.py]: short comment

**Comment by Eric89GXL** [comment]
Check out the tests, hopefully those will make us confident :) If they don't, then they can be expanded.

File: mne/bem.py (hunk 2 above)
Comment (Eric89GXL) [inline mne/bem.py]: That would require (re-)understanding what it does first :) but seriously, I'll add one. It also reminds me that I need to check for duplication of functionality across the `_compute_forward` module, I think these are different but I should check.

File: mne/bem.py [hunk 3]
```diff
@@ -12,10 +12,512 @@
...
+def _fwd_bem_multi_solution(solids, gamma, nps):
+    """Do multi surface solution
+
+      * Invert I - solids/(2*M_PI)
+      * Take deflation into account
+      * The matrix is destroyed after inversion
+      * This is the general multilayer case
+
+    """
+    from scipy import linalg
```
Comment (dengemann) [inline mne/bem.py]: why import here?

Comment (Eric89GXL) [inline mne/bem.py]: we nest certain scipy imports now to save on import speed, see `mne/tests/test_import_nesting.py`

Comment (dengemann) [inline mne/bem.py]: right! thanks

File: mne/bem.py [hunk 4]
```diff
@@ -447,6 +955,356 @@ def make_watershed_bem(subject, subjects_dir=None, overwrite=False,
...
+
+    res['rr'] = tag.data.astype(np.float)  # XXX : double because of mayavi bug
+    if res['rr'].shape[0] != res['np']:
+        raise ValueError('Vertex information is incorrect')
+
+    tag = find_tag(fid, this, FIFF.FIFF_MNE_SOURCE_SPACE_NORMALS)
+    if tag is None:
+        tag = tag = find_tag(fid, this, FIFF.FIFF_BEM_SURF_NORMALS)
+    if tag is None:
+        res['nn'] = []
```
Comment (dengemann) [inline mne/bem.py]: you have a couple of these. list() would be more canonical.

Comment (Eric89GXL) [inline mne/bem.py]: I think these lines were copy-pasted, but I'll change them to conform

**Comment by Eric89GXL** [comment]
@dengemann comments addressed. FYI you asked about speed, and that made me optimize a bit more:
```
2 (320)  1       C       0.283627033234
2 (320)  1       Python  0.291443109512
2 (320)  3       C       0.465950012207
2 (320)  3       Python  2.45959687233
3 (1280)  1       C       0.660896062851
3 (1280)  1       Python  1.62492084503
3 (1280)  3       C       12.4721641541
3 (1280)  3       Python  11.3650319576
4 (5120)  1       C       24.9352688789
4 (5120)  1       Python  11.2639200687
4 (5120)  3       C       644.138396025
4 (5120)  3       Python  118.89906311
```

**Comment by dengemann** [comment]
Great!!! I will recompute all my BEMs to enjoy the speedy drive :)

File: mne/bem.py [hunk 5]
```diff
@@ -12,10 +12,518 @@
...
+    """The linear potential matrix element computations"""
+    from .source_space import _fast_cross_nd_sum
+    omega = np.zeros((len(fros), 3))
+
+    # we replicate a little bit of the _get_solids code here for speed
+    v1 = tri_rr[np.newaxis, 0, :] - fros
+    v2 = tri_rr[np.newaxis, 1, :] - fros
+    v3 = tri_rr[np.newaxis, 2, :] - fros
+    triples = _fast_cross_nd_sum(v1, v2, v3)
+    l1 = np.sqrt(np.sum(v1 * v1, axis=1))
```
Comment (agramfort) [inline mne/bem.py]: is it faster than np.linalg.norm( ..., axis=1) ?

Comment (Eric89GXL) [inline mne/bem.py]: Yes sadly (and `axis` is only available on 1.9+), I tested both. I think it depends on your application. Probably for larger matrices `norm(..., axis)` would be faster, we'll have to decide on a case-by-case basis.

**Comment by agramfort** [comment]
please update what's new, rebase and +1 for merge !

thanks awesome !

**Comment by Eric89GXL** [comment]
Rebased and updated, I'll merge once Travis comes back happy

**Comment by agramfort** [comment]
Sounds good

**Comment by Eric89GXL** [comment]
Thanks for the reviews @dengemann @agramfort

**Comment by dengemann** [comment]
Super cool, Eric -- amazing!

**Pull request merged**